File: library/src/main/java/com/handy/basic/utils/ActivityStackUtils.java
package com.handy.basic.utils;

import android.app.Activity;

import java.util.Collections;
import java.util.Iterator;
import java.util.Stack;

/**
 * Activity栈工具类
 *
 * @author LiuJie https://github.com/Handy045
 * @description functional description.
 * @date Created in 2019/2/27 16:53
 * @modified By liujie
 */
public final class ActivityStackUtils {

    /**
     * Activity栈
     */
    private static Stack<Activity> activityStack;

    private ActivityStackUtils() {
        throw new UnsupportedOperationException("u can't instantiate me...");
    }

    /**
     * 获取当前Activity（堆栈中最后一个压入的）
     */
    public static Activity getCurrent() {
        if (activityStack != null) {
            return activityStack.lastElement();
        }
        return null;
    }

    /**
     * 添加Activity到堆栈
     */
    public static void addActivity(Activity activity) {
        if (activityStack == null) {
            activityStack = new Stack<>();
        }
        activityStack.add(activity);
    }

    /**
     * 结束当前Activity（堆栈中最后一个压入的）
     */
    public static void finish() {
        if (activityStack != null) {
            Activity activity = activityStack.lastElement();
            finishChoiceDesc(activity);
        }
    }

    /**
     * 按顺序单次结束指定的Activity
     */
    public static void finishChoiceAsc(Activity activity) {
        if (activity != null && activityStack != null) {
            activityStack.remove(activity);
            if (!activity.isFinishing()) {
                activity.finish();
            }
        }
    }

    /**
     * 按倒序单次结束指定的Activity
     */
    public static void finishChoiceDesc(Activity activity) {
        if (activity != null && activityStack != null) {
            /*倒序排列*/
            Collections.reverse(activityStack);
            activityStack.remove(activity);
            if (!activity.isFinishing()) {
                activity.finish();
            }
            /*排列还原*/
            Collections.reverse(activityStack);
        }
    }

    /**
     * 结束全部指定的Activit
     */
    public static void finishChoices(Activity activity) {
        if (activity != null && activityStack != null) {
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                if (aty.getClass().equals(activity.getClass())) {
                    iterator.remove();
                    if (!aty.isFinishing()) {
                        aty.finish();
                    }
                }
            }
        }
    }

    /**
     * 按顺序单次结束指定的Activity
     */
    public static void finishChoiceAsc(Class<?> cls) {
        if (activityStack != null) {
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                if (aty.getClass().equals(cls)) {
                    iterator.remove();
                    if (!aty.isFinishing()) {
                        aty.finish();
                    }
                    break;
                }
            }
        }
    }

    /**
     * 按倒序单次结束指定的Activity
     */
    public static void finishChoiceDesc(Class<?> cls) {
        if (activityStack != null) {
            /*倒序排列*/
            Collections.reverse(activityStack);
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                if (aty.getClass().equals(cls)) {
                    iterator.remove();
                    if (!aty.isFinishing()) {
                        aty.finish();
                    }
                    break;
                }
            }
            /*排列还原*/
            Collections.reverse(activityStack);
        }
    }

    /**
     * 结束全部指定的Activit
     */
    public static void finishChoices(Class<?> cls) {
        if (activityStack != null) {
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                if (aty.getClass().equals(cls)) {
                    iterator.remove();
                    if (!aty.isFinishing()) {
                        aty.finish();
                    }
                }
            }
        }
    }

    /**
     * 结束所有Activity
     */
    public static void finishAll() {
        if (activityStack != null) {
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                iterator.remove();
                if (!aty.isFinishing()) {
                    aty.finish();
                }
            }
            activityStack.clear();
        }
    }

    /**
     * 结束第一个Activity之后的所有Activity
     */
    public static void finish2First() {
        if (activityStack != null) {
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                if (activityStack.size() > 1) {
                    iterator.remove();
                    if (!aty.isFinishing()) {
                        aty.finish();
                    }
                }
            }
        }
    }

    /**
     * 按顺序结束到指定的Activity
     */
    public static void finish2ChoiceAsc(Class<?> cls) {
        if (activityStack != null) {
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                if (!aty.getClass().equals(cls)) {
                    iterator.remove();
                    if (!aty.isFinishing()) {
                        aty.finish();
                    }
                } else {
                    break;
                }
            }
        }
    }

    /**
     * 按倒序结束到指定的Activity
     */
    public static void finish2ChoiceDesc(Class<?> cls) {
        if (activityStack != null) {
            /*倒序排列*/
            Collections.reverse(activityStack);
            Iterator iterator = activityStack.iterator();
            while (iterator.hasNext()) {
                Activity aty = (Activity) iterator.next();
                if (!aty.getClass().equals(cls)) {
                    iterator.remove();
                    if (!aty.isFinishing()) {
                        aty.finish();
                    }
                } else {
                    break;
                }
            }
            /*排列还原*/
            Collections.reverse(activityStack);
        }
    }

    public static Stack<Activity> getActivityStack() {
        return activityStack;
    }

    public static void setActivityStack(Stack<Activity> activityStack) {
        ActivityStackUtils.activityStack = activityStack;
    }
}

File: library/src/main/java/com/handy/basic/utils/ActivityStackUtils.kt
package com.handy.basic.utils

import android.app.Activity
import java.util.*

/**
 * @title: ActivityStackUtils
 * @projectName HandyBasicKT
 * @description: Activity堆栈手动管理工具类
 * @author LiuJie https://github.com/Handy045
 * @date Created in 2019-11-01 15:30
 */
class ActivityStackUtils private constructor() {
    companion object {
        private var activityStack: Stack<Activity> = Stack()

        /**
         * 获取当前Activity
         */
        fun getCurrent(): Activity? {
            return if (!activityStack.isNullOrEmpty()) activityStack.lastElement() else null
        }

        /**
         * 结束当前Activity
         */
        fun finishCurrent() {
            if (!activityStack.isNullOrEmpty()) {
                Collections.reverse(activityStack)
                val activity = activityStack.lastElement()
                activityStack.remove(activity)
                if (!activity.isFinishing) {
                    activity.finish()
                }
                Collections.reverse(activityStack)
            }
        }

        /**
         * 重载当前Activity
         */
        fun reloadCurrent() {
            if (!activityStack.isNullOrEmpty()) {
                val activity = activityStack.lastElement()
                val intent = activity.intent
                activity.finish()
                activity.startActivity(intent)
            }
        }

        /**
         * 添加一个Activity
         */
        fun addActivity(activity: Activity) {
            if (activityStack.isNullOrEmpty()) {
                activityStack = Stack()
            }
            activityStack.add(activity)
        }

        /**
         * 倒序中单次结束指定的Activity
         */
        fun finishChoiceDesc(activity: Activity) {
            if (!activityStack.isNullOrEmpty()) {
                Collections.reverse(activityStack)
                finishChoiceAsc(activity)
                Collections.reverse(activityStack)
            }
        }

        /**
         * 正序中单次结束指定的Activity
         */
        fun finishChoiceAsc(activity: Activity) {
            if (!activityStack.isNullOrEmpty()) {
                activityStack.remove(activity)
                if (!activity.isFinishing) {
                    activity.finish()
                }
            }
        }

        /**
         * 倒序中移除到指定的Activity
         */
        fun finishToChoiceDesc(activity: Activity) {
            if (!activityStack.isNullOrEmpty()) {
                Collections.reverse(activityStack)
                finishToChoiceAsc(activity)
                Collections.reverse(activityStack)
            }
        }

        /**
         * 升序中移除到指定的Activity
         */
        fun finishToChoiceAsc(activity: Activity) {
            if (!activityStack.isNullOrEmpty()) {
                val iterator = activityStack.iterator()
                while (iterator.hasNext()) {
                    val aty = iterator.next()
                    if (aty.javaClass != activity.javaClass) {
                        iterator.remove()
                        if (!aty.isFinishing) {
                            aty.finish()
                        }
                    } else {
                        break
                    }
                }
            }
        }

        /**
         * 移除所有指定的Activity
         */
        fun finishChoiceAll(activity: Activity) {
            if (!activityStack.isNullOrEmpty()) {
                val iterator = activityStack.iterator()
                while (iterator.hasNext()) {
                    val aty = iterator.next()
                    if (aty.javaClass == activity.javaClass) {
                        iterator.remove()
                        if (!aty.isFinishing) {
                            aty.finish()
                        }
                    }
                }
            }
        }

        /**
         * 移除到第一个Activity
         */
        fun finishToFirst() {
            if (!activityStack.isNullOrEmpty()) {
                val iterator = activityStack.iterator()
                while (iterator.hasNext()) {
                    val aty = iterator.next()
                    if (activityStack.size > 1) {
                        iterator.remove()
                        if (!aty.isFinishing) {
                            aty.finish()
                        }
                    }
                }
            }
        }

        /**
         * 移除全部的Activity
         */
        fun finishAll() {
            if (!activityStack.isNullOrEmpty()) {
                val iterator = activityStack.iterator()
                while (iterator.hasNext()) {
                    val aty = iterator.next()
                    iterator.remove()
                    if (!aty.isFinishing) {
                        aty.finish()
                    }
                }
                activityStack.clear()
            }
        }
    }
}